### PR TITLE
Support additional extraResources types from Patient/$everything

### PR DESCRIPTION
**Pull request by lmsurpre**
Instead of a big if/else with all the manually preconfigured search parameters, we now look up the applicable search parameters for the given resource type for this tenant and then process those into the set of includes.

It should work mostly the same, except that:
1. we now support "includeTypes" for all resource types; previously it only supported "Medication", "Location", "Practitioner", and "Organization
2. for the 4 types that were previously supported, we include a few additional includes that were missed in the manually computed list
3. we fix a single issue with an include value from ExplanationOfBenefit (it used to add "insurer" but this is not a valid search parameter in the core specification)

### CODE REVIEW

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     */
+    public static List<String> computeIncludesParamValues(String baseResourceType, List<String> resourceTypesToInclude, SearchHelper searchHelper)
+            throws Exception {
+        List<String> result = new ArrayList<>();
+
+        List<String> allowedIncludes = FHIRConfigHelper.getSearchPropertyRestrictions(baseResourceType, FHIRConfigHelper.SEARCH_PROPERTY_TYPE_INCLUDE);
+        Map<ResourceType, List<String>> codesByType = IncludesUtil.getSearchCodesByType(baseResourceType, searchHelper);
+
+        for (Map.Entry<ResourceType, List<String>> entry : codesByType.entrySet()) {
+            if (resourceTypesToInclude.contains(entry.getKey().value())) {
```
Comment (punktilious) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:42]: Multiple calls to List.contains here. Would perform better if resourceTypesToInclude were a Set, which also makes more sense from a semantic point of view too.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                continue;
+            }
+
+            String expression = sp.getExpression().getValue();
+            for (ResourceTypeCode target : sp.getTarget()) {
+                // The search parameter target types come from the possible types of the target reference
+                // and don't take into account any "is X" filters from within the expression,
+                // so we use this heuristic to avoid adding the code when expression filtering applies
+                // and does not include a filter clause for this particular target type.
+                if (expression.contains(".is(") || expression.contains(" is ")) {
```
Comment (punktilious) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:72]: Is this brittle?

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            List<String> allowedIncludes, List<String> codes) throws Exception {
+        List<String> includes = new ArrayList<>();
+
+        for (String code : codes) {
+            // Need to make sure the search parameter has not been excluded
+            String parameterName = compartmentMemberType + ":" + code + ":" + subResourceType.value();
+            String simplifiedParameterName = compartmentMemberType + ":" + code;
+
+            boolean isAllowed = allowedIncludes == null ||
+                    allowedIncludes.contains(parameterName) ||
```
Comment (punktilious) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:104]: Calling List#contains within a loop. Might be more appropriate to use a Set<String> here.

File: operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java [hunk 4]
```diff
@@ -765,8 +457,7 @@ private void readsOfAdditionalAssociatedResources(String compartmentMemberType,
                     String type = referenceValue.getTargetResourceType();
                     String value = referenceValue.getValue();
                     String resourceId = type + "/" + value;
-                    if (!resourceIds.contains(resourceId) && extraResources.contains(type) && (type.equals(ResourceType.LOCATION.value()) || type.equals(ResourceType.MEDICATION.value())
-                            || type.equals(ResourceType.ORGANIZATION.value()) || type.equals(ResourceType.PRACTITIONER.value()))) {
+                    if (!resourceIds.contains(resourceId) && extraResources.contains(type)) {
```
Comment (punktilious) [inline operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:460]: Call to List#contains. Slow.

File: operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java [hunk 5]
```diff
@@ -744,12 +434,14 @@ private void addSearchParameterIfNotExcluded(String compartmentMemberType, Strin
      * @param compartmentMemberType the type of resource currently drilling down on
      * @param newEntries a list of Bundle response entries that include compartmentMemberType instances found via search
      * @param allEntries a list of all entries that have been found so far (minus duplicates)
+     * @param resourceIds the set of external resource ids ("resourceType/id") to be included in the result (to avoid duplicates)
      * @param resourceHelper an instance of FHIRResourceHelpers to use to make the calls
      * @param extraResources a list of "extra" resource types configured for this server
```
Comment (punktilious) [inline operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:439]: I think extraResourceTypes would be a better parameter name

File: operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java [hunk 6]
```diff
@@ -807,7 +498,7 @@ private void readsOfAdditionalAssociatedResources(String compartmentMemberType,
      * @param retrieveAdditionalResources whether the server config has specified to retrieve additional resources
      */
     private void processResults(Bundle results, String compartmentMemberType, List<Entry> newEntries,
-                List<Entry> allEntries, List<String> resourceIds, FHIRResourceHelpers resourceHelper, List<String> extraResources,
+                List<Entry> allEntries, Set<String> resourceIds, FHIRResourceHelpers resourceHelper, List<String> extraResources,
```
Comment (punktilious) [inline operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:501]: Perhaps `Set<String> extraResourceTypes`

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:72]: its definitely "heuristic".  IMHO the conformance artifacts should account for this when they list the allowed "target" types for a search parameter of type reference.

Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:72]: a potential alternative would be for us to patch the official resources in order to make those right...then we wouldn't need to do this.  but it wouldn't necessarily help with user-packaged IGs that may have the same issue

Comment (punktilious) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:72]: I don't think it's a blocker...just struck me that the test could hit false positives...but I don't know enough about the context.

Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:72]: ok.  i think its pretty safe.  theoretically someone could construct a FHIRPath to confuse us (like by combining an “all” or “iff” or “not” or something like that with these “is” clauses), but in practice no one does that for search parameters.

Its almost always just a simple path and sometimes they have an is at the end to filter to a particular type.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java (hunk 1 above)
Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java:43]: done

File: operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java (hunk 4 above)
Comment (lmsurpre) [reply operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java]: done

File: operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java (hunk 6 above)
Comment (lmsurpre) [reply operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java]: done

**Pull request merged**